uuid/v7.ts

```diff
@@ -0,0 +1,95 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+import { bytesToUuid } from "./_common.ts";
+
+const UUID_RE =
+  /^[0-9a-f]{8}-[0-9a-f]{4}-[7][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
+/**
+ * Determines whether a string is a valid
+ * {@link https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7 | UUIDv7}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @param id UUID value.
+ *
+ * @returns `true` if the string is a valid UUIDv7, otherwise `false`.
+ *
+ * @example Usage
+ * ```ts
+ * import { validate } from "@std/uuid/v7";
+ * import { assert, assertFalse } from "@std/assert";
+ *
+ * assert(validate("017f22e2-79b0-7cc3-98c4-dc0c0c07398f"));
+ * assertFalse(validate("fac8c1e0-ad1a-4204-a0d0-8126ae84495d"));
+ * ```
+ */
+export function validate(id: string): boolean {
+  return UUID_RE.test(id);
+}
+
+/**
+ * Options for {@linkcode generate}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ */
+export interface GenerateOptions {
+  /**
+   * Unix epoch timestamp in milliseconds.
+   *
+   * @default {Date.now()}
+   */
+  timestamp?: number;
+  /**
+   * 10 bytes of random value to use in the UUID.
+   * Generally you should not need to set these, but it can be useful for testing.
+   *
+   * @internal
+   */
+  random?: Uint8Array;
+}
+
+const version = 0b0111;
+const variant = 0b10;
+
+/**
+ * Generates a {@link https://www.rfc-editor.org/rfc/rfc9562.html#section-5.7 | UUIDv7}.
+ *
+ * @experimental **UNSTABLE**: New API, yet to be vetted.
+ *
+ * @param options Options for generating the UUIDv7.
+ * @returns Returns a UUIDv7 string
+ *
+ * @example Usage
+ * ```ts
+ * import { generate, validate } from "@std/uuid/v7";
+ * import { assert } from "@std/assert";
+ *
+ * const uuid = generate();
+ * assert(validate(uuid));
+ * ```
+ */
+export function generate(options: GenerateOptions = {}): string {
+  const bytes = new Uint8Array(16);
+  const view = new DataView(bytes.buffer);
+  // Unix timestamp in milliseconds (truncated to 48 bits)
+  if (options.timestamp !== undefined) {
+    if (!Number.isInteger(options.timestamp) || options.timestamp < 0) {
+      throw new Error(
+        `Cannot generate UUID as timestamp must be a non-negative integer: timestamp ${options.timestamp}`,
+      );
+    }
+  }
+  const timestamp = BigInt(options.timestamp ?? Date.now());
+  view.setBigUint64(0, timestamp << 16n);
```

Comment:
Why must we deal with a `BigInt` here? Genuine question.

Reply:
Convenience only, `BigInt` is not required.

Alternatives could be something along the lines of:
```ts
const shiftedTimestamp = (timestamp & 0xFFFFFFFFFFFF) << 16;
view.setUint32(0, shiftedTimestamp / 0x100000000); // High 32 bits
view.setUint32(2, shiftedTimestamp & 0xFFFFFFFF); // Low 32 bits
```
or
```
data[0] = (timestamp / 0x10000000000) & 0xff;
data[1] = (timestamp / 0x100000000) & 0xff;
data[2] = (timestamp / 0x1000000) & 0xff;
data[3] = (timestamp / 0x10000) & 0xff;
data[4] = (timestamp / 0x100) & 0xff;
data[5] = timestamp & 0xff;
```

Would need test the difference in performance

Reply:
Ok. Let's keep as is then.

Reply:
To clarify, the simplicity of what we have now looks good. But if there's a sufficient performance difference, then we should re-consider.

Reply:
Thanks for the clarification. I will do some naive local testing tonight to check if performance is significantly different and report back.

Reply:
I wrote naive benchmarks on [this branch](https://github.com/Liam-Tait/std/tree/liam/uuid-v7-view-array-benchmark)

[bench](https://github.com/Liam-Tait/std/blob/34e42926b11bf931be2c334aa7f0d6f9b5284b47/uuid/v7_bench.ts)
[changes](https://github.com/Liam-Tait/std/blob/74edfcc02affcb7d5a2a709324fd3daa2e0e0672/uuid/v7.ts)

I am not getting results that gives me confidence either way. Sometimes faster, sometimes slower. Let me know if you get something more consistent.
Happy to go one way or the other.

Reply:
Ok. Let's stick with what we have now then.

Reply:
According to my brief local benchmark, your implementation seems to be ~1.5x-2x faster than `npm:uuid`, but I'm not sure if other machines will yield the same results.